doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,214 @@
+
+## Motivation
+SR-IOV Network Operator configures SR-IOV one node at a time and one nic at a same time. That means we’ll need to wait
+hours or even days to configure all NICs  on large cluster deployments. Also moving all draining logic to a centralized
+place which will reduce chances of race conditions and bugs that were encountered before in sriov-network-config-daemon
+with draining.
+
+### Use Cases
+
+### Goals
+* Number of drainable nodes should be 1 by default
+* Number of drainable nodes should be configured by pool
+* Nodes pool should be defined by node selector
+* A node could be included into several pools. In this case pool should be 
+  selected by it’s priority like it’s implemented for SriovNetworkNodePolicy
+* Move all drain-related logic into the centralized place
+
+
+### Non-Goals
+Parallel NICs configuration on the same node is out of scope of this proposal
+
+## Proposal
+Introduce nodes pool drain configuration and controller to meet Goals targets.
+
+
+### Workflow Description
+A new Drain controller will be introduced to manage node drain and cordon procedures. That means we don't need to do
+drain and use `drain lock` in config daemon anymore. The overall drain process will be covered by the following states:
+
+```golang
+NodeDrainAnnotation             = "sriovnetwork.openshift.io/state"
+NodeStateDrainAnnotation        = "sriovnetwork.openshift.io/desired-state"
+NodeStateDrainAnnotationCurrent = "sriovnetwork.openshift.io/current-state"
+DrainIdle                       = "Idle"
+DrainRequired                   = "Drain_Required"
+RebootRequired                  = "Reboot_Required"
+Draining                        = "Draining"
+DrainComplete                   = "DrainComplete"
+```
+
+Drain controller will watch for node annotation, `sriovnetwork.openshift.io/state` 
+and SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state`
+and write the `sriovnetwork.openshift.io/current-state` annotation in the SriovNetworkNodeState.
+
+Config Daemon will read SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state` and write both
+node annotation, `sriovnetwork.openshift.io/state` and SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state`.
+
+*NOTE:* In the future we are going to drop the node annotation and only use the SriovNetworkNodeState
+
+Draining procedure:
+
+1. config daemon mark the node as `Drain_Required` or `Reboot_Required`
+2. operator drain controller reconcile loop find the node SriovNetworkPoolConfig
+   1. if number of `Draining` nodes is great or equal to the `MaxUnavailable` the operator will re-queue the request
+   2. if number of `Draining` nodes is lower than the `MaxUnavailable` the operator will start the draining process
+5. on Openshift platform we will pause the machine config pool related to the node
+6. the operator will start the drain process
+   1. if `Drain_Required` the operator will remove ONLY pods used sriov devices
```

Comment:
Thats not really related to parallel sriov config.

Reply:
but it's part of the changes do you want me to remove it?

Reply:
yes please, and preferably address it in a separate PR on top if it makes sense.

we should either remove from design and implementation or keep in both. i prefer the former and a folllowup PR to fine-tune the draining process.

Reply:
if possible I will like to leave it here as the  implementation is already in place and the change is super small we just check if we have `Drain_required` or `Reboot_requried` but if you insist I can remove it :)

Reply:
as its already there and both the design and implementation already went through several iterations lets keep it.